### DISCUSSION TITLE
adding components to the canvas method

### DISCUSSION THREAD

**Discussion by 2524268884** [mui/toolpad#2782] [Help and support]
I have developed a form container component myself and bound it to the API of Google Sheet. Through this API, I can obtain the type and name of each field in Google Sheet. Next, I hope that within the form container, I can automatically add corresponding form components (I have developed a set of form components myself) based on the field types returned to me by the API. If I want to achieve this effect, does it mean that ToolPad has relevant JS methods that can be called? For example, adding components to the canvas?

**Comment by prakhargupta1** [comment]
Hi, @2524268884 While I'll let someone from the engineering team reply to this, I wanted to [invite](https://calendly.com/prakhar-mui/toolpad?back=1&month=2023-10) you for a user interview to understand your use case better. It will help us unblock you by getting a better idea of the missing features. If you are available, we can have a 30-minute discussion this week. Thanks

**Reply by 2524268884** [reply]
Hi, thank you for your reply. I would love to communicate with you through a meeting, but my English is not very good. Therefore, if we communicate during the meeting, it may be difficult to chat. I think I can try my best to use translation software to express my needs during the meeting. At the same time, I will also share the screen to express some of the functional scenarios I currently want. If you think this is okay, I think we can communicate online. I will schedule a meeting time first and look forward to your reply

**Reply by prakhargupta1** [reply]
Hi, Invite received. We can communicate however you are comfortable. I have also invited @Janpot. Talk to you soon!

**Comment by bharatkashyap** [comment]
While Toolpad doesn't have a way to programmatically add components to the canvas, you can programmatically control the display state of a component already added to the canvas through bindings. For instance, you could do something like this by binding to the `sx` property of the component:

https://github.com/mui/mui-toolpad/assets/19550456/307db061-2202-4ab8-8ad9-de87842fa992

**Comment by prakhargupta1** [comment]
Hi @2524268884 We have a related [feature request](https://github.com/mui/mui-toolpad/issues/1770.). Would this suffice your use case?
